sdk/couchbase/src/clients/cluster_client.rs

```diff
@@ -0,0 +1,169 @@
+                let bucket_client = backend.bucket(name).await?;
+
+                Ok(BucketClient::new(
+                    BucketClientBackend::Couchbase2BucketBackend(bucket_client),
+                ))
+            }
+        }
+    }
+
+    pub fn query_client(&self) -> error::Result<QueryClient> {
+        match &self.backend {
+            ClusterClientBackend::CouchbaseClusterBackend(backend) => {
+                let query_client = backend.query_client()?;
+
+                Ok(QueryClient::new(
+                    QueryClientBackend::CouchbaseQueryClientBackend(query_client),
+                ))
+            }
+            ClusterClientBackend::Couchbase2ClusterBackend(_) => {
+                unimplemented!()
+            }
+        }
+    }
+}
+
+struct CouchbaseClusterBackend {
+    agent_manager: Arc<OnDemandAgentManager>,
+}
+
+impl CouchbaseClusterBackend {
+    pub async fn connect(
+        hosts: Vec<Address>,
+        _srv_record: Option<SrvRecord>,
+        use_ssl: bool,
+        opts: ClusterOptions,
+        extra_opts: HashMap<String, Vec<String>>,
+    ) -> error::Result<CouchbaseClusterBackend> {
+        let mut opts: OnDemandAgentManagerOptions = opts.into();
+
+        if opts.tls_config.is_some() && !use_ssl {
+            return Err(error::Error {
+                msg: "TLS config provided but couchbase scheme used".into(),
+            });
+        } else if opts.tls_config.is_none() && use_ssl {
+            return Err(error::Error {
+                msg: "No TLS config provided but couchbases scheme used".into(),
+            });
+        }
+
+        Self::merge_options(&mut opts, extra_opts)?;
+
+        opts.seed_config.memd_addrs = hosts.iter().map(|a| a.to_string()).collect();
+
+        let mgr = OnDemandAgentManager::new(opts).await?;
+
+        Ok(Self {
+            agent_manager: Arc::new(mgr),
+        })
+    }
+
+    async fn bucket(&self, name: String) -> error::Result<CouchbaseBucketClient> {
+        let agent = self.agent_manager.get_bucket_agent(&name).await?;
+
+        Ok(CouchbaseBucketClient::new(agent, name))
+    }
+
+    fn query_client(&self) -> error::Result<CouchbaseQueryClient> {
+        let agent = self.agent_manager.get_cluster_agent()?;
+
+        Ok(CouchbaseQueryClient::new(agent.clone()))
+    }
+
+    fn merge_options(
+        opts: &mut OnDemandAgentManagerOptions,
+        extra_opts: HashMap<String, Vec<String>>,
+    ) -> error::Result<()> {
+        for (k, v) in extra_opts {
+            match k.as_str() {
+                "kv_connect_timeout" => {
+                    opts.connect_timeout =
+                        Some(Duration::from_millis(v[0].parse().map_err(|e| {
+                            error::Error {
+                                msg: format!("{}", e),
+                            }
+                        })?))
+                }
+                "placeholder" => {}
+                _ => (),
```

Comment:
Is it worth logging if extra_opts are being ignored?

Reply:
Maybe, but I've logged an issue to expand our parsing - for now this is POC.